core/chains/terra/types/chain.go

```diff
@@ -0,0 +1,222 @@
+package types
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink/core/utils"
+
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+
+	"github.com/smartcontractkit/sqlx"
+
+	"github.com/smartcontractkit/chainlink-terra/pkg/terra"
+	terraclient "github.com/smartcontractkit/chainlink-terra/pkg/terra/client"
+	terraconfig "github.com/smartcontractkit/chainlink-terra/pkg/terra/config"
+
+	"github.com/smartcontractkit/chainlink/core/chains/terra/terratxm"
+	"github.com/smartcontractkit/chainlink/core/config"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/keystore"
+	"github.com/smartcontractkit/chainlink/core/services/pg"
+)
+
+// ChainSetOpts holds options for configuring a ChainSet.
+type ChainSetOpts struct {
+	Config           config.GeneralConfig
+	Logger           logger.Logger
+	DB               *sqlx.DB
+	KeyStore         keystore.Terra
+	EventBroadcaster pg.EventBroadcaster
+	ORM              ORM
+}
+
+var _ terra.ChainSet = (*chainSet)(nil)
+
+type chainSet struct {
+	opts   ChainSetOpts
+	mu     sync.RWMutex
+	chains map[string]terra.Chain
+	lggr   logger.Logger
+}
+
+// NewChainSet returns a new chain set for opts.
+func NewChainSet(opts ChainSetOpts) (terra.ChainSet, error) {
+	dbchains, err := opts.ORM.EnabledChainsWithNodes()
+	if err != nil {
+		return nil, errors.Wrap(err, "error loading chains")
+	}
+	cs := &chainSet{
+		opts:   opts,
+		chains: make(map[string]terra.Chain),
+		lggr:   opts.Logger.Named("ChainSet"),
+	}
+	for _, c := range dbchains {
+		n := c.Nodes[0] //TODO client pool
+		var err2 error
+		cs.chains[c.ID], err2 = NewChain(opts.DB, opts.KeyStore, n, opts.Config, opts.EventBroadcaster, c.Cfg, opts.Logger)
+		if err2 != nil {
+			err = multierr.Combine(err, err2)
+			continue
+		}
+	}
+	return cs, err
+}
+
+func (c *chainSet) Get(id string) (terra.Chain, error) {
+	c.mu.RLock()
+	ch := c.chains[id]
+	c.mu.RUnlock()
+	if ch != nil {
+		return ch, nil
+	}
+	c.mu.Lock()
+	defer c.mu.Unlock()
```

Comment:
no RLock / RUnlock?

Reply:
We need the write lock here since we are (trying to) update the map.

Reply:
it `Get()` really the best name then? feels strange to me to be doing state changes in a getter function.

Reply:
Originally inherited from the EVM version (which doesn't have any adhoc), but I think it could still be interpreted as a Get (based on _DB_ state) with the internal caching and lazy load being an implementation detail. This will continue to evolve as we fill in chain configuration and node pool pieces as well.

Reply:
Renamed to `Chain()` which is aligned with other interfaces, and doesn't imply anything about implementation.